lib/cretonne/meta/isa/intel/legalize.py

```diff
@@ -53,6 +53,15 @@
                 (a, dead) << x86.sdivmodx(x, xhi, y)
             ))
 
+
+#intel_expand.legalize(
+#        a << insts.func_addr.bind(i64)(x),
+#        Rtl(
+#            # First, get the pointer-to-x using a 32-bit
+#            # rip-relative mov; and reloc with REX_GOTPCRELX 
+#            # Second, get x from that pointer (ordinary load)
```

Comment:
This sounds right, except for the second load. The GOT holds the address of x, which is what `func_addr` needs to compute.

The other thing to watch out for here is the difference between `R_X86_64_GOTPCREL` and `R_X86_64_GOTPCRELX`. The latter is an optimized variant for the case where you know you're not using a REX prefix (and `R_X86_64_REX_GOTPCRELX` is for when you know you *are* using a REX prefix). The easier thing to do here would be to just use plain `R_X86_64_GOTPCREL` and leave a note to revisit this when desired.

Reply:
Hey there. I'm not totally sure I follow what you're saying re: the second load. 

```
$ gcc -O0 -fPIC -c pic2.c -o pic2.o
$ cat pic2.c
extern void (*foo)();

int main() {
	void (*bar) = foo;
	return 0;
}

$ objdump -dx pic2.o

... snip ... 

Disassembly of section .text:

0000000000000000 <main>:
   0:	55                   	push   %rbp
   1:	48 89 e5             	mov    %rsp,%rbp
   4:	48 8b 05 00 00 00 00 	mov    0x0(%rip),%rax        # b <main+0xb>
			7: R_X86_64_GOTPCREL	foo-0x4
   b:	48 8b 00             	mov    (%rax),%rax
   e:	48 89 45 f8          	mov    %rax,-0x8(%rbp)
  12:	b8 00 00 00 00       	mov    $0x0,%eax
  17:	5d                   	pop    %rbp
  18:	c3                   	retq
```

It looks to me like these two lines imply a two-step process here when loading a pointer from the GOT:
```
   4:	48 8b 05 00 00 00 00 	mov    0x0(%rip),%rax        # b <main+0xb>
			7: R_X86_64_GOTPCREL	foo-0x4
   b:	48 8b 00             	mov    (%rax),%rax
```

Reply:
Compare that with what you get when you change this:
```c
extern void (*foo)();
```
to this:
```
extern void foo();
```
In the first form, the GOT load gets the address of foo, which is a global variable holding the address of some function. In the second, the GOT load gets the address of foo, which is a function. The second is what `func_addr` does.

Reply:
Oh jeez. Thank you, @sunfishcode.